### PR TITLE
refactor ReleaseMain and AsyncDownloader to prevent multiple download requests

### PR DESCRIPTION
**Pull request by snopoke**
http://manage.dimagi.com/default.asp?185771
Turns out that every time you click download on the releases page a new JS object would be created and listen the 'show' event on the modal. After 3 downloads you'd have 3 download objects listening and sending off ajax requests for the download.

If the objects were created from different download URL's (ccz for v1, ccz for v2, multimedia zip for v3) then the actual file downloaded would be the request that finished last.

@benrudolph 
@sravfeyn

### CODE REVIEW

File: corehq/apps/app_manager/static/app_manager/spec/releases_spec.js [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+        function get_saved_apps(num) {
+            return _.map(_.range(num), function (version) {
+                return {
+                    id: version,
+                    version: version,
+                    doc_type: 'Application',
+                    build_comment: '',
+                    build_broken: false,
+                    is_released: false
+                }
```
Comment (dimagimon) [inline corehq/apps/app_manager/static/app_manager/spec/releases_spec.js]: Missing semicolon.

**Comment by czue** [comment]
26 commits for an interrupt bug. nice

**Comment by snopoke** [comment]
@calellowitz FYI

**Comment by snopoke** [comment]
:white

**Comment by proteusvacuum** [comment]
Nice! I think I introduced this in one of my first HQ PR's. :cry: 
Also, javascript tests!

File: .gitignore [hunk 2]
```diff
@@ -53,3 +53,4 @@ celerybeat-schedule
 # compiled translations
 /locale/*/*/*.mo
 .grunt
+node_modules/
```
Comment (benrudolph) [inline .gitignore]: lol saw this and had no idea how i could've missed this, but i think i have node_modules in my global ignore. good catch

File: Gruntfile.js [hunk 3]
```diff
@@ -25,7 +25,10 @@ module.exports = function(grunt) {
         mochaConfig[app] = {
             options: {
                 urls: [BASE_URL + appName + '/' + (config ? config : '')],
-                run: true
+                run: true,
+                log: true,
+                logErrors: true,
+                reporter: 'Spec'
```
Comment (benrudolph) [inline Gruntfile.js]: :smile:

File: corehq/apps/app_manager/static/app_manager/js/releases.js [hunk 4]
```diff
@@ -1,6 +1,5 @@
-function SavedApp(o, r) {
-    var $root = r;
-    var self = ko.mapping.fromJS(o);
+function SavedApp(app_data, releasesMain) {
```
Comment (benrudolph) [inline corehq/apps/app_manager/static/app_manager/js/releases.js]: yess hate the single letter variable names

**Comment by benrudolph** [comment]
:+1: one really minor thing i see is the underscore notations, i think in general javascript is always written in camelCase (all major libs). we don't have a style guide for javascript but would be nice to adopt one one day. one that i adhere to in the absence of one is airbnb's ES5 [styleguide](https://github.com/airbnb/javascript/tree/master/es5)

**Comment by snopoke** [comment]
Good point on style. I tried to stick to whatever convention was being used in the JS file I was editing. Perhaps I'll do a style prefactor next time

**Pull request merged**